### ISSUE TITLE
Linter doesn't handle comments within the `RUN` shell block

### ISSUE THREAD

**Issue by dnwe** [hadolint/hadolint#76]
Having comments with a `RUN` block are perfectly valid shell, and `docker build` handles them without issue.

e.g., (contrived example)
```
$ cat Dockerfile
FROM alpine:edge

RUN apk --no-cache add zsh \
  # now add vim
  && apk --no-cache add vim
```

```
% docker build -t temp . 
Sending build context to Docker daemon 2.048 kB
Step 1/2 : FROM alpine:edge
 ---> 3fc33d6d5e74
Step 2/2 : RUN apk --no-cache add zsh   && apk --no-cache add vim
 ---> Running in 842e35506e47
fetch http://dl-cdn.alpinelinux.org/alpine/edge/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/edge/community/x86_64/APKINDEX.tar.gz
(1/4) Installing ncurses-terminfo-base (6.0-r7)
(2/4) Installing ncurses-terminfo (6.0-r7)
(3/4) Installing ncurses-libs (6.0-r7)
(4/4) Installing zsh (5.3.1-r0)
Executing zsh-5.3.1-r0.post-install
Executing busybox-1.24.2-r9.trigger
OK: 17 MiB in 15 packages
fetch http://dl-cdn.alpinelinux.org/alpine/edge/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/edge/community/x86_64/APKINDEX.tar.gz
(1/2) Installing lua5.2-libs (5.2.4-r2)
(2/2) Installing vim (8.0.0187-r0)
Executing busybox-1.24.2-r9.trigger
OK: 43 MiB in 17 packages
 ---> 760d383ada47
Removing intermediate container 842e35506e47
Successfully built 760d383ada47
```

However, hadolint currently fails to handle them:
```
% docker run --rm -i lukasmartinelli/hadolint <Dockerfile
/dev/stdin:5:1 unexpected '&' expecting space, "\t", "ONBUILD", "FROM", "COPY", "RUN", "WORKDIR",
"ENTRYPOINT", "VOLUME", "EXPOSE", "ENV", "ARG", "USER", "LABEL", "STOPSIGNAL", "CMD", "SHELL",
"MAINTAINER", "ADD", "#", "HEALTHCHECK" or end of input
```

Unless you strip them out with sed:
```
% sed -e '/\s\+#/d' < Dockerfile | docker run --rm -i lukasmartinelli/hadolint 
/dev/stdin DL4000 Specify a maintainer of the Dockerfile
```

**Comment by vergenzt** [comment]
Yeah came here to post the same issue:

```sh
$ cat Dockerfile
FROM busybox:latest

RUN echo "line one" \
 # comment
 && echo "line two"
```
```sh
$ docker build .
Sending build context to Docker daemon 2.048 kB
Step 1/2 : FROM busybox:latest
 ---> 00f017a8c2a6
Step 2/2 : RUN echo "line one"  && echo "line three"
 ---> Running in 11a9f704773a
line one
line three
 ---> bc4b904b6b3b
Removing intermediate container 11a9f704773a
Successfully built bc4b904b6b3b
```
```sh
$ hadolint Dockerfile
Dockerfile:5:1 unexpected '&' expecting space, "\t", "ONBUILD", "FROM", "COPY", "RUN", "WORKDIR", "ENTRYPOINT", "VOLUME", "EXPOSE", "ENV", "ARG", "USER", "LABEL", "STOPSIGNAL", "CMD", "SHELL", "MAINTAINER", "ADD", "#", "HEALTHCHECK" or end of input
```
```sh
$ hadolint -v
Haskell Dockerfile Linter v1.2.1
```

**Comment by zemanlx** [comment]
@dnwe @vergenzt you can solve it by escaping newline for comment and you are good to go or at least it works for me.

From 

```bash
RUN echo "line one" \
 # comment
 && echo "line two"
```

to

```bash
RUN echo "line one" \
 # comment \
 && echo "line two"
```

**Comment by zemanlx** [comment]
@lorenzo I believe we should solve this in the parser as they are `dockerfile` comments and not `shell` comments. What do you think?
